source/mir/base64.d

```diff
@@ -0,0 +1,317 @@
+        shift = cast(ubyte)('0' - 52);
+    }
+    else if (i == 62)
+    {
+        // character plus
+        shift = cast(ubyte)('+' - 62);
+    }
+    else if (i == 63)
+    {
+        // character slash
+        shift = cast(ubyte)('/' - 63);
+    }
+
+    return cast(char)(i + shift);
+}
+
+// Do the inverse of above (convert an ASCII value into the Base64 character set)
+private ubyte lookup_decoding(char i) @safe @nogc pure
+{
+    // Branching bad, but this isn't performance sensitive
+    if (i <= 'Z' && i >= 'A') {
+        return cast(ubyte)(i - 'A');
+    }
+    else if (i <= 'z' && i >= 'a') {
+        return cast(ubyte)(i - 'a' + 26); 
+    }
+    else if (i <= '9' && i >= '0') {
+        return cast(ubyte)(i - '0' + 52);
+    }
+    else if (i == '+') {
+        return 62;
+    }
+    else if (i == '/') {
+        return 63;
+    }
+    // Just return 0 for padding,
+    // as it typically means nothing.
+    else if (i == '=') {
+        return 0;
+    }
+    else {
+        version(D_Exceptions) {
+            throw base64DecodeInvalidCharException;
+        } else {
+            assert(0, base64DecodeInvalidCharMsg);
+        }
+    }
+
+}
+
+/++
+Decode a Base64 encoded value, returning the buffer.
++/
+ubyte[] decodeBase64(scope ubyte[] buf) @safe pure
+{
+    import mir.appender : scopedBuffer;
+    auto app = scopedBuffer!ubyte;
+    decodeBase64(buf, app);
+    return app.data.dup;
+}
+
+/++
+Decode a Base64 encoded value, placing the result onto an Appender.
++/
+void decodeBase64(Appender)(scope ubyte[] input, ref Appender appender) @safe @nogc pure
+{
+    // We expect data should be well-formed (with padding),
+    // so we should throw if it is not well-formed.
+    if (input.length % 4 != 0)
+    {
+        version(D_Exceptions) {
+            throw base64DecodeInvalidLenException;
+        } else {
+            assert(0, base64DecodeInvalidLenMsg);
+        }
+    }
+    foreach(group; input.bytegroup!(4, uint).map!bswap)
+    {
+        // We only expect valid ASCII values for these,
+        // hence the 0x7f.
+        const(ubyte) a = lookup_decoding((group >> 24) & 0x7f);
+        const(ubyte) b = lookup_decoding((group >> 16) & 0x7f);
+        const(ubyte) c = lookup_decoding((group >> 8) & 0x7f);
+        const(ubyte) d = lookup_decoding((group) & 0x7f);
+
+        // We do the inverse of how we encoded it...
+        uint transformed_group = (a << 26) | (b << 20) | (c << 14) | (d << 8);
+
+        const(ubyte) t_a = (transformed_group >> 24) & 0xff;
+        const(ubyte) t_b = (transformed_group >> 16) & 0xff;
+        const(ubyte) t_c = (transformed_group >> 8) & 0xff;
+        const(ubyte) t_d = (transformed_group) & 0xff;
+
+        // We should *always* have enough for at least
+        // one, but we don't need to have enough for the rest..
+        appender.put(t_a);
+
+        // Only emit transformed groups if we have enough data for them.
+        if (t_b == 0 && t_c == 0 && t_d == 0)
```

Comment:
weird and verbose condition branching. Will `AAAA` pass?

Reply:
It actually turns out that my logic was faulty here -- `t_d` would always be zero (which I think of at the time of writing).